File: DGEMM/SemOtimizacao_Paralelismo/dgemmSparalelismo.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

/*
 Função para multiplicar duas matrizes quadradas A e B,
 armazenando o resultado em C.
 Implementação direta sem paralelismo, usando dgemm clássico, sem otimizações de BLAS.
*/
void dgemm(int n, double *A, double *B, double *C) {
    for (int i = 0; i < n; i++) {                               // Percorre as linhas de A
        for (int j = 0; j < n; j++) {                           // Percorre as colunas de B
            C[i * n + j] = 0.0;                                 // Inicializa elemento C[i][j]
            for (int k = 0; k < n; k++) {                       // Produto escalar da linha i (A) e coluna j (B)
                C[i * n + j] += A[i * n + k] * B[k * n + j];
            }
        }
    }
}

/*
 Função para gerar uma matriz de dimensão (rows x cols)
 preenchida com valores aleatórios no intervalo [0, 10].
*/
void generateRandomMatrix(int rows, int cols, double *matrix) {
    for (int i = 0; i < rows * cols; i++) {                     // Percorre cada elemento da matriz p gerar:  
        int randomInt = rand();                                 // Inteiro aleatório entre 0 e RAND_MAX
        double randomDouble = (double)randomInt / RAND_MAX;     // Normaliza para [0,1] (convertendo int p/ double)
        matrix[i] = randomDouble * 10.0;                        // Escalar o valor para [0,10]
    }
}

int main() {
    // Inicializa a semente do gerador de números aleatórios com o tempo atual
    srand(time(NULL));

    // Conjunto de tamanhos de matrizes a serem testados
    int sizes[] = {256, 512, 1024, 2048, 4096, 8192};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);

    // Executa 3 rodadas (run = 6, 7 e 8)
    for (int run = 6; run <= 8; run++) {
        char filename[20];
        snprintf(filename, sizeof(filename), "matrizes_c_%d.txt", run);

        printf("Execução %d:\n", run);

        // Abre o arquivo de saída para salvar tempos de execução
        FILE *file = fopen(filename, "w");
        if (file == NULL) {
            perror("Erro ao abrir o arquivo para escrita");
            return 1;
        }

        // Testa cada tamanho definido em 'sizes'
        for (int i = 0; i < num_sizes; i++) {
            int n = sizes[i];   // Tamanho da matriz (n x n)

            // Aloca dinamicamente memória para A, B e C
            double *A = (double *)malloc(n * n * sizeof(double));
            double *B = (double *)malloc(n * n * sizeof(double));
            double *C = (double *)malloc(n * n * sizeof(double));

            // Preenche A e B com valores aleatórios
            generateRandomMatrix(n, n, A);
            generateRandomMatrix(n, n, B);

            // Marca tempo inicial
            clock_t start = clock();

            // Realiza a multiplicação A * B → C
            dgemm(n, A, B, C);

            // Marca tempo final
            clock_t end = clock();

            // Calcula tempo decorrido em segundos
            double elapsed_time = (double)(end - start) / CLOCKS_PER_SEC;

            // Exibe no console
            printf("Matrix size: %dx%d, Execution time: %.4lf seconds\n",
                   n, n, elapsed_time);

            // Grava o tempo no arquivo
            fprintf(file, "Matrix size: %dx%d, Execution time: %.4lf seconds\n",
                    n, n, elapsed_time);

            // Libera a memória alocada para as matrizes
            free(A);
            free(B);
            free(C);
        }

        // Fecha o arquivo
        fclose(file);
        printf("Arquivo '%s' salvo com sucesso!\n", filename);
        printf("\n");
    }

    return 0;
}

File: DGEMM/SemOtimizacao_Paralelismo/dgemmSparalelismo.py
import random
import time
import sys

# ------------------------------------------------------------------------------
# Função: gerar_matriz
# Objetivo: cria uma matriz (lista de listas em Python) de dimensão
#           num_linhas × num_colunas, preenchida com números inteiros aleatórios
#           no intervalo [1, 10].
# Observação: a geração é feita linha a linha, utilizando a função randint
#             da biblioteca random.
# Complexidade temporal: O(num_linhas · num_colunas)
# ------------------------------------------------------------------------------
def gerar_matriz(num_linhas, num_colunas):
    matriz = []
    for _ in range(num_linhas):
        # Cria uma linha da matriz contendo 'num_colunas' elementos aleatórios
        linha = [random.randint(1, 10) for _ in range(num_colunas)]
        matriz.append(linha)
    return matriz


# ------------------------------------------------------------------------------
# Função: multiplicar_matrizes
# Objetivo: realiza a multiplicação matricial clássica (sem paralelismo)
#           entre duas matrizes (matriz1 × matriz2).
#
# Pré-condição: o número de colunas da matriz1 deve ser igual ao número
#               de linhas da matriz2 (dimensões compatíveis).
# Pós-condição: retorna uma nova matriz contendo o resultado da multiplicação.
#
# Fórmula matemática: 
#   Se C = A × B, então:
#       C[i][j] = Σ ( A[i][k] · B[k][j] ), k = 0..p-1
#
# Complexidade temporal: O(n³) para matrizes quadradas de ordem n.
# Complexidade espacial: O(n²) para armazenar a matriz resultado.
# ------------------------------------------------------------------------------
def multiplicar_matrizes(matriz1, matriz2):
    # Obtém dimensões de ambas as matrizes
    linhas_matriz1 = len(matriz1)
    colunas_matriz1 = len(matriz1[0])
    linhas_matriz2 = len(matriz2)
    colunas_matriz2 = len(matriz2[0])

    # Verificação de compatibilidade: condição fundamental para multiplicação
    if colunas_matriz1 != linhas_matriz2:
        raise ValueError("O número de colunas da matriz1 deve ser igual "
                         "ao número de linhas da matriz2")

    # Inicializa a matriz resultado com zeros (dimensão: linhas_matriz1 × colunas_matriz2)
    resultado = [[0 for _ in range(colunas_matriz2)] for _ in range(linhas_matriz1)]

    # Estrutura de laços triplos: varredura completa da operação matricial
    for i in range(linhas_matriz1):              # percorre as linhas da matriz1
        for j in range(colunas_matriz2):         # percorre as colunas da matriz2
            for k in range(colunas_matriz1):     # percorre os elementos comuns
                # Produto elemento a elemento e acumulação da soma
                resultado[i][j] += matriz1[i][k] * matriz2[k][j]

    return resultado


# ------------------------------------------------------------------------------
# Programa principal
# Objetivo: executar múltiplas rodadas da multiplicação de matrizes de dimensão
#           8192 × 8192, registrando o tempo de execução em arquivos .txt.
#
# Observação: matrizes 8192×8192 consomem memória em ordem de dezenas de GB e
#             execução extremamente lenta em Python puro (sem NumPy).
#             O objetivo aqui é acadêmico (didático), não prático.
# ------------------------------------------------------------------------------
for i in range(2, 8):
    # Cria um arquivo de saída distinto para cada iteração
    with open(f"matriz_python_{i}.txt", "w") as arquivo:
        # Redireciona a saída padrão (stdout) para o arquivo
        original_stdout = sys.stdout
        sys.stdout = arquivo

        # Texto isolado (não é impresso porque não há 'print')
        "Calculo da matriz 8192 x 8192"

        # Geração das matrizes A e B
        matriz_A = gerar_matriz(8192, 8192)
        matriz_B = gerar_matriz(8192, 8192)

        # Marca tempo inicial
        inicio = time.time()

        # Realiza a multiplicação matricial
        matriz_C = multiplicar_matrizes(matriz_A, matriz_B)

        # Marca tempo final
        fim = time.time()

        # Calcula e registra o tempo de execução
        print(f"Execution time for 8192 x 8192 matrix: {round(fim - inicio, 3)} seconds")

        # Restaura a saída padrão original
        sys.stdout = original_stdout